src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt

```diff
@@ -1396,32 +1430,300 @@ abstract class ManagedStrategy(
+         * they have the same ids produced by [MethodIds] and the same parameters. That means we don't care
+         * here where this method is called - it's more convenient to see as short version of the spin cycle as possible.
+         *
+         * But this approach may produce a wrong result in case when we switch from the spin lock and then occur
+         * in the same spin lock again.
+         * Let's consider this situation using the example above.
+         * Trace points of the second spin lock would have the following call stacks:
+         *
+         * 1. `actor -> [a -> b -> c] -> [a -> b -> c]`
+         * 2. `actor -> [a -> b -> c] -> [a -> b -> c]`
+         *
+         * And the first execution point of the second iteration will be
+         * ```
+         *     actor -> [a -> b -> c] -> [a -> b -> c] -> [a -> b -> c]
+         * ```
+         *
+         * If we apply this algorithm to the first point from the first iteration and the first point from the second iteration
+         * we get `[a -> b -> c] -> [a -> b -> c]` longest common suffix, which is not correct, because the first recursive call
+         * of this cycle starts at the depth `actor -> a -> b`.
+         * In this case, the last trace point before the spin cycle becomes handy.
+         * In the example above it would have the following call stack `actor -> a -> b -> c`.
+         *
+         * Here are the points we're interested in:
+         * 1. `actor -> [a -> b -> c]` - The last trace point before the cycle.
+         * 2. `actor -> [a -> b -> c] -> [a -> b -> c]` - The first trace point of the cycle.
+         * 3. ` actor -> [a -> b -> c] -> [a -> b -> c] -> [a -> b -> c]` - The first trace point of the second iteration.
+         * Let's note the following rule: if point 1 has the same method call on the same depth as point 2,
+         * then this call can't be a part of the current spin cycle.
+         * The reason is quite easy: if it was, we would mark point 1 as a first spin cycle trace point.
+         *
+         * So, to detect the spin cycle, start trace point depth in case of a recursive spin lock we:
+         * 1. Take the first trace point of the spin cycle on the first iteration (point A) and on the second iteration (point B).
+         * 2. Take the last trace point before the spin cycle (point C).
+         * 3. We walk synchronously on their call stacks from the end to the beginning, while call from the A stack
+         * is the same as a call from the stack B. But it also equals the call from the call C, then we need to stop.
+         * The count of the call passed the condition above equals to the call we need to lift from the first spin cycle
+         * node to get the correct spin cycle start trace point depth.
+         *
+         * @param beforeMethodCallSwitch flag if this method invoked right after [MethodCallTracePoint] is added to a trace,
+         * before the corresponding method is called.
+         */
```

Comment:
Because all this logic is related to the spin-lock detection, can we consider moving at least part of it out of the `TraceCollector` into `LoopDetector.kt` ?